harness/determined/launch/autodeepspeed.py

```diff
@@ -0,0 +1,256 @@
+    assert container_id is not None, "Unable to run with DET_CONTAINER_ID unset"
+
+    # TODO: refactor websocket, data_layer, and profiling to to not use the cli_cert.
+    cert = certs.default_load(info.master_url)
+    certs.cli_cert = cert
+
+    # The launch layer should provide the chief_ip to the training code, so that the training code
+    # can function with a different launch layer in a different environment.  Inside Determined, the
+    # easiest way to get the chief_ip is with container_addrs.
+    chief_ip = info.container_addrs[0]
+
+    # Chief IP is set as an environment variable to support nested launch layers
+    os.environ["DET_CHIEF_IP"] = chief_ip
+
+
+    # All ranks will need to run sshd.
+    run_sshd_command = [
+        "/usr/sbin/sshd",
+        "-p",
+        str(constants.DTRAIN_SSH_PORT),
+        "-f",
+        "/run/determined/ssh/sshd_config",
+        "-D",
+    ]
+
+    if info.container_rank > 0:
+        # Non-chief machines just run sshd.
+
+        # Mark sshd containers as daemon containers that the master should kill when all non-daemon
+        # containers (deepspeed launcher, in this case) have exited.
+        api.post(
+            info.master_url,
+            path=f"/api/v1/allocations/{info.allocation_id}/containers/{container_id}/daemon",
+            cert=cert,
+        )
+
+        # Wrap it in a pid_server to ensure that we can't hang if a worker fails.
+        # This is useful for deepspeed which does not have good error handling for remote processes
+        # spun up by pdsh.
+        pid_server_cmd = [
+            "python3",
+            "-m",
+            "determined.exec.pid_server",
+            "--on-fail",
+            "SIGTERM",
+            "--on-exit",
+            "SIGTERM",
+            "--grace-period",
+            "3",
+            f"/tmp/pid_server-{info.allocation_id}",
+            str(len(info.slot_ids)),
+            "--",
+        ]
+
+        logging.debug(
+            f"Non-chief [{info.container_rank}] training process launch "
+            f"command: {run_sshd_command}."
+        )
+        return subprocess.Popen(pid_server_cmd + run_sshd_command).wait()
+
+    # We always need to set this variable to initialize the context correctly, even in the single
+    # slot case.
+    os.environ["USE_DEEPSPEED"] = "1"
+
+    if len(info.container_addrs) > 1:
+        # Create the environment file that will be passed by deepspeed to individual ranks.
+        create_deepspeed_env_file()
+        # Set custom PDSH args:
+        # * bypass strict host checking
+        # * -p our custom port
+        # * other args are default ssh args for pdsh
+        os.environ["PDSH_SSH_ARGS"] = (
+            "-o PasswordAuthentication=no -o StrictHostKeyChecking=no "
+            f"-p {constants.DTRAIN_SSH_PORT} -2 -a -x %h"
+        )
+
+        # Chief worker also needs to run sshd when using pdsh.
+        subprocess.Popen(run_sshd_command)
```

Comment:
This process is still never being cleaned up.  You need some try/finally mechanics here.